qiita_db/metadata_template.py

```diff
@@ -1445,6 +1462,89 @@ def extend(self, md_template):
         # adding the fp to the object
         self.add_filepath(fp)
 
+    def update(self, md_template):
+        r"""Update values in the sample metadata template
+
+        Parameters
+        ----------
+        md_template : DataFrame
+            The metadata template file contents indexed by samples Ids
+
+        Raises
+        ------
+        QiitaDBError
+            If md_template and db do not have the same sample ids
+            If md_template and db do not have the same column headers
+        """
+        conn_handler = SQLConnectionHandler()
+
+        # Clean and validate the metadata template given
+        new_map = self._clean_validate_template(md_template, self.id,
+                                                conn_handler)
+        # Retrieving current metadata
+        current_map = self._transform_to_dict(conn_handler.execute_fetchall(
+            "SELECT * FROM qiita.{0} WHERE {1}=%s".format(self._table,
+                                                          self._id_column),
+            (self.id,)))
+        dyn_vals = self._transform_to_dict(conn_handler.execute_fetchall(
+            "SELECT * FROM qiita.{0}".format(self._table_name(self.id))))
+
+        for k in current_map:
+            current_map[k].update(dyn_vals[k])
+            try:
```

Comment:
Ah ... I see what you are saying. From searching on the intertubes, it
seems like this is a perfect use-case for dict.pop, for example:

```
In [1]: x = {'a': 1, 'b': 2, 'c': 3}

In [2]: x.pop('a', None)
Out[2]: 1

In [3]: x.pop('LOLERCOASTER', None)
```


On (Mar-06-15|11:01), Antonio Gonzalez wrote:
> > +        conn_handler = SQLConnectionHandler()
> > +
> > +        # Clean and validate the metadata template given
> > +        new_map = self._clean_validate_template(md_template, self.id,
> > +                                                conn_handler)
> > +        # Retrieving current metadata
> > +        current_map = self._transform_to_dict(conn_handler.execute_fetchall(
> > +            "SELECT * FROM qiita.{0} WHERE {1}=%s".format(self._table,
> > +                                                          self._id_column),
> > +            (self.id,)))
> > +        dyn_vals = self._transform_to_dict(conn_handler.execute_fetchall(
> > +            "SELECT * FROM qiita.{0}".format(self._table_name(self.id))))
> > +
> > +        for k in current_map:
> > +            current_map[k].update(dyn_vals[k])
> > +            try:
> 
> Wouldn't that make it slower? This way we are not searching within a list every single time
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/qiita/pull/904/files#r25968099